### ISSUE TITLE
Automatically write new defaults and indexes in migrations

### ISSUE THREAD

**Issue by a-type** [a-type/verdant#182]
lo-fi knows better than you do what got added, so there's not much reason to ask the user to do this.

For perf we could track which changed collections have already been run through `migrate` and not redo those.

**Comment by a-type** [comment]
This is evolving into a bigger effort on updating the way schemas and migrations work.

Key goals:

1. Make migrations as seamless and automated as possible - reduce code to write
2. Since migrations live on in the codebase indefinitely... they should be as small as possible, and maybe abstract enough to modify underlying implementation in new library releases without requiring code changes
3. Local-first carries more risk if local data on user devices is compromised in some way. Migrations should include as much validation and advance warning of potential issues as possible before runtime
4. Schemas should be modular and allow reuse of sub-schemas, maybe even recursive definitions


1 & 2 kind of go together. The idea is making migrations simple, do as much obvious work for the user as possible internally, and focus on higher-level APIs which won't get too bound to internal implementation. Ideally the underlying systems can iterate without forcing users to rewrite all their historical migrations.

Another possibility is doing 'migration collapse' -- when a user determines that certain schema versions are no longer in use, allowing them to define a new migration which covers a whole version range.

3: Still thinking about this. What it does imply is that I probably don't want to be picky about validating actual entity objects against the schema at runtime. Encouraging defensive coding around looser schema contracts is more resilient to unexpected failures vs. encouraging reliance on stricter schema definitions. For that reason any musings about making schemas a Zod validator are kind of out the window. The big question is "what should happen if the validation fails at runtime on a user's device?" and I don't see any satisfying answer to it.

However, it's not clear where the line should be drawn on validation. I liked the idea of creating migrations which are statically typechecked as the key validation. This kind of fell apart as the internal typings of the library became too complex and I had to fall back to generated types. Generated types for migrations is an option, but where will those (probably large) generated typings live? On the flip side, these typings would never change for historical versions, in theory. They could also utilize named types exported from the library itself which could be updated if mistakes were made, even if those types are implemented as basically pass-throughs initially. Embedding these types in the migration files (maybe at the bottom) would be convenient as the literal historical schemas are no longer needed, unblocking my "interpreted CLI" work I had to shelve since transpiling the schemas and storing them was feeling too fragile.

4 would be unblocked by the interpreted CLI work as well.

**Comment by a-type** [comment]
Migration ideation...

```ts
export default migration(
	v3Migration,
	async ({ convert }) => {
		// generated code, you can replace with your own migration procedure
		await convert.items((old) => {
			// TODO: implement the conversion! Return the new shape for an "items" entity. Summary of changes:
 			// - required "name" field added with no default
			// "tags[number]" shape changed from string to object
		});
	},
);

/** WARNING: this metadata enforces the correctness of your migration. only modify it if you're confident Verdant has made a mistake in determining migration requirements. Modifying this to avoid fulfilling requirements could result in corrupted and irrepairable data on user devices. */
const v3Migration = migrationInfo<{
	collections: {
		items: [
			{ id: string; tags: string[]; value?: string; }
			{ id: string; tags: { name: string; color: string; }[]; value?: string; name: string; }
		],
	},
}>({ required: ['items'] });
```

If the `convert.[name]` invocations are removed, the migration should throw an error on startup. However, since the code and typings are all in userland, the user is empowered to fix any problems or mistakes.

**Comment by a-type** [comment]
Initial prototypes indicate historical schemas are still relevant. Doesn't seem like there's much way around it.

I think I may have to just get comfortable with the idea of storing historical schema modules in the user's codebase.

**Comment by a-type** [comment]
Yeah, this is looking more and more like a no-op for any major changes to migrations themselves, but continuing on the path of a CLI which interprets the schema.
